### ISSUE TITLE
Cannot create Invoices and contacts

### ISSUE THREAD

**Issue by sksankarraj** [XeroAPI/xero-node#508]
**SDK:**
 - Version - 4.11.0

**Describe the bug**
I'm trying to create Invoices and Contacts using Xero-Node library. But facing the same behaviour in both. Here I explained what I did for Contacts

**To Reproduce**
```
  const contacts = {
    contacts: [
      {
        'name': data.first_name + ' ' + data.last_name,
        'firstName': data.first_name,
        'lastName': data.last_name,
        'emailAddress': data.email,
        'addresses': [
          {
            'addressType': 'STREET',
            'addressLine1': data.address.replace('$$$$', ', '),
            'city': data.city,
            'postalCode': data.zip_code
          }
        ],
      }
    ],
  };
  console.log(contacts);
  try {
    const response = await xero.accountingApi.createContacts(tokenSet.tenant_id, contacts, summarizeErrors);
    logger.info('Invoice created', response.body || response.response.statusCode);
    return response;
  } catch (err) {
    const error = JSON.stringify(err, null, 2);
    logger.error(`Status Code: ${err} => ${error}`);
  }
}; 
```

Output:

```
// from console.log
{
  contacts: [
    {
      name: 'patient name3',
      firstName: 'patient',
      lastName: 'name3',
      emailAddress: 'name@patient.com',
      addresses: [Array]
    }
  ]
}

// error
"response": {
    "statusCode": 400,
    "body": {
      "ErrorNumber": 10,
      "Type": "ValidationException",
      "Message": "A validation exception occurred",
      "Elements": [
        {
          "ContactID": "00000000-0000-0000-0000-000000000000",
          "Name": "",
          "Addresses": [],
          "Phones": [],
          "ContactGroups": [],
          "ContactPersons": [],
          "HasValidationErrors": true,
          "ValidationErrors": [
            {
              "Message": "Contact name cannot be empty"
            }
          ]
        }
      ]
    },
    "headers": {
      "content-type": "application/json; charset=utf-8",
      "content-length": "475",
      "server": "nginx",
      "x-appminlimit-remaining": "9995",
      "x-minlimit-remaining": "58",
      "x-daylimit-remaining": "4859",
      "expires": "Sun, 21 Mar 2021 05:03:00 GMT",
      "cache-control": "max-age=0, no-cache, no-store",
      "pragma": "no-cache",
      "date": "Sun, 21 Mar 2021 05:03:00 GMT",
      "connection": "close",
      "x-client-tls-ver": "tls1.3"
    },
    "request": {
      "uri": {
        "protocol": "https:",
        "slashes": true,
        "auth": null,
        "host": "api.xero.com",
        "port": null,
        "hostname": "api.xero.com",
        "hash": null,
        "search": "?summarizeErrors=true",
        "query": "summarizeErrors=true",
        "pathname": "/api.xro/2.0/Contacts",
        "path": "/api.xro/2.0/Contacts?summarizeErrors=true",
        "href": "https://api.xero.com/api.xro/2.0/Contacts?summarizeErrors=true"
      },
      "method": "PUT",
      "headers": {
        "user-agent": "xero-node-4.11.0",
        "xero-tenant-id": <tenant-id>,
        "Authorization": "Bearer  token",
        "accept": "application/json",
        "content-type": "application/json",
        "content-length": 231
      }
    }
  },
  "body": {}
}
```

So I `console.log`ged here - https://github.com/XeroAPI/xero-node/blob/master/src/gen/api/accountingApi.ts#L1314
```
console.log(models_1.ObjectSerializer.serialize(contacts, "Contacts"))
```
I got this as input data.
```{
  Contacts: [
    {
      ContactID: undefined,
      ContactNumber: undefined,
      AccountNumber: undefined,
      ContactStatus: undefined,
      Name: 'patient name3',
      FirstName: 'patient',
      LastName: 'name3',
      EmailAddress: 'name@patient.com',
      SkypeUserName: undefined,
      ContactPersons: undefined,
      BankAccountDetails: undefined,
      TaxNumber: undefined,
      AccountsReceivableTaxType: undefined,
      AccountsPayableTaxType: undefined,
      Addresses: [Array],
      Phones: undefined,
      IsSupplier: undefined,
      IsCustomer: undefined,
      DefaultCurrency: undefined,
      XeroNetworkKey: undefined,
      SalesDefaultAccountCode: undefined,
      PurchasesDefaultAccountCode: undefined,
      SalesTrackingCategories: undefined,
      PurchasesTrackingCategories: undefined,
      TrackingCategoryName: undefined,
      TrackingCategoryOption: undefined,
      PaymentTerms: undefined,
      UpdatedDateUTC: undefined,
      ContactGroups: undefined,
      Website: undefined,
      BrandingTheme: undefined,
      BatchPayments: undefined,
      Discount: undefined,
      Balances: undefined,
      Attachments: undefined,
      HasAttachments: undefined,
      ValidationErrors: undefined,
      HasValidationErrors: undefined,
      StatusAttributeString: undefined,
      undefined: undefined
    },
    {
      ContactID: undefined,
      ContactNumber: undefined,
      AccountNumber: undefined,
      ContactStatus: undefined,
      Name: '',
      FirstName: undefined,
      LastName: undefined,
      EmailAddress: undefined,
      SkypeUserName: undefined,
      ContactPersons: undefined,
      BankAccountDetails: undefined,
      TaxNumber: undefined,
      AccountsReceivableTaxType: undefined,
      AccountsPayableTaxType: undefined,
      Addresses: undefined,
      Phones: undefined,
      IsSupplier: undefined,
      IsCustomer: undefined,
      DefaultCurrency: undefined,
      XeroNetworkKey: undefined,
      SalesDefaultAccountCode: undefined,
      PurchasesDefaultAccountCode: undefined,
      SalesTrackingCategories: undefined,
      PurchasesTrackingCategories: undefined,
      TrackingCategoryName: undefined,
      TrackingCategoryOption: undefined,
      PaymentTerms: undefined,
      UpdatedDateUTC: undefined,
      ContactGroups: undefined,
      Website: undefined,
      BrandingTheme: undefined,
      BatchPayments: undefined,
      Discount: undefined,
      Balances: undefined,
      Attachments: undefined,
      HasAttachments: undefined,
      ValidationErrors: undefined,
      HasValidationErrors: undefined,
      StatusAttributeString: undefined,
      undefined: undefined
    }
  ],
  undefined: undefined
}
```

So it seems like the serialization adds one empty object in the array. Am I missing anything here? Or is it a Bug?

Thanks in advance.

**Comment by RettBehrens** [comment]
Hi @sksankarraj. I just tried to replicate your issue and I only console log one object in the serializer as expected. It looks like you are sending a contact array with two objects - the second object with only a name property set as an empty string? What is logged if you console.log(contacts.length) just before calling the API? Closing this for now as it does not appear to be an SDK issue.

**Comment by sksankarraj** [comment]
As You can see in the code(that's exactly what i'm running), its only one contact. I even console.loged it and got one for `console.log(contacts.length)`. Its very strange that it throws that error.

I'm using JavaScript not TypeScript, Can you please provide me a sample code for JS, in the docs I can only see the TS code.

**Comment by sksankarraj** [comment]
> Hi @sksankarraj. I just tried to replicate your issue and I only console log one object in the serializer as expected. It looks like you are sending a contact array with two objects - the second object with only a name property set as an empty string? What is logged if you console.log(contacts.length) just before calling the API? Closing this for now as it does not appear to be an SDK issue.

As You can see in the code(that's exactly what i'm running), its only one contact. I even console.loged it and got one for `console.log(contacts.length)`. Its very strange that it throws that error.

I'm using JavaScript not TypeScript, Can you please provide me a sample code for JS, in the docs I can only see the TS code.

**Comment by RettBehrens** [comment]
Can you send me a link to your repo?

**Comment by thrivedigital** [comment]
Was this ever actually resolved? I've been having the same issue, and with the Oauth change deadline coming up at the end of the month, it's preventing us from updating our internal app. Thanks!

**Comment by sksankarraj** [comment]
@thrivedigital , No it never actually fixed. Issue is still there in the latest version. I actually didn't waste time on this. So I went back to using traditional http calls.

**Comment by thrivedigital** [comment]
@sksankarraj Thank you for the heads up, I guess I will just have to try making HTTP calls with axios too...

**Comment by RettBehrens** [comment]
I am unable to replicate the issue using Xero-Node v4.15.0. 

With each release that adds new methods we add those methods to our kitchen sink sample app and test those calls to verify functionality.
https://github.com/XeroAPI/xero-node-oauth2-app

In the past 30 days Xero-Node users have made:
POST: 4,963,388 API calls
PUT: 2,159,183 API calls

I've tested using two of our sample apps:

Node and Express without TypeScript
https://github.com/XeroAPI/xero-node-serializer-test

Node and Express with TypeScript
https://github.com/XeroAPI/xero-node-oauth2-ts-starter

```
serializer call:  {
  Invoices: [
    {
      Type: 'ACCREC',
      Contact: [Object],
      LineItems: [Array],
      Date: '2021-09-24',
      DueDate: '2021-09-25',
      LineAmountTypes: undefined,
      InvoiceNumber: undefined,
      Reference: undefined,
      BrandingThemeID: undefined,
      Url: undefined,
      CurrencyCode: undefined,
      CurrencyRate: undefined,
      Status: undefined,
      SentToContact: undefined,
      ExpectedPaymentDate: undefined,
      PlannedPaymentDate: undefined,
      CISDeduction: undefined,
      CISRate: undefined,
      SubTotal: undefined,
      TotalTax: undefined,
      Total: undefined,
      TotalDiscount: undefined,
      InvoiceID: undefined,
      RepeatingInvoiceID: undefined,
      HasAttachments: undefined,
      IsDiscounted: undefined,
      Payments: undefined,
      Prepayments: undefined,
      Overpayments: undefined,
      AmountDue: undefined,
      AmountPaid: undefined,
      FullyPaidOnDate: undefined,
      AmountCredited: undefined,
      UpdatedDateUTC: undefined,
      CreditNotes: undefined,
      Attachments: undefined,
      HasErrors: undefined,
      StatusAttributeString: undefined,
      ValidationErrors: undefined,
      Warnings: undefined
    }
  ]
}
localVarRequestOptions:  {
  method: 'PUT',
  qs: {},
  headers: {
    'user-agent': 'xero-node-4.15.0',
    'xero-tenant-id': '07f32319-daaf-4548-b589-d6f06f7b407f'
  },
  uri: 'https://api.xero.com/api.xro/2.0/Invoices',
  useQuerystring: false,
  json: true,
  body: { Invoices: [ [Object] ] }
}
invoices:  [
  Invoice {
    type: 'ACCREC',
    contact: Contact {
      contactID: 'adf801de-a142-4670-906f-0d46c5d7c693',
      contactNumber: undefined,
      accountNumber: undefined,
      contactStatus: 'ACTIVE',
      name: 'Gable Print',
      firstName: undefined,
      lastName: undefined,
      emailAddress: '',
      skypeUserName: undefined,
      contactPersons: [],
      bankAccountDetails: '',
      taxNumber: undefined,
      accountsReceivableTaxType: undefined,
      accountsPayableTaxType: undefined,
      addresses: [Array],
      phones: [Array],
      isSupplier: false,
      isCustomer: true,
      defaultCurrency: undefined,
      xeroNetworkKey: undefined,
      salesDefaultAccountCode: undefined,
      purchasesDefaultAccountCode: undefined,
      salesTrackingCategories: [],
      purchasesTrackingCategories: [],
      trackingCategoryName: undefined,
      trackingCategoryOption: undefined,
      paymentTerms: undefined,
      updatedDateUTC: 2021-09-13T18:03:42.080Z,
      contactGroups: [],
      website: undefined,
      brandingTheme: undefined,
      batchPayments: undefined,
      discount: undefined,
      balances: undefined,
      attachments: undefined,
      hasAttachments: undefined,
      validationErrors: undefined,
      hasValidationErrors: false,
      statusAttributeString: undefined
    },
    lineItems: [ [LineItem] ],
    date: 2021-09-24T00:00:00.000Z,
    dueDate: 2021-09-25T00:00:00.000Z,
    lineAmountTypes: 'Exclusive',
    invoiceNumber: 'INV-0055',
    reference: '',
    brandingThemeID: '5d4dd402-c851-497e-aae1-9ff265c0d15a',
    url: undefined,
    currencyCode: 'GBP',
    currencyRate: 1,
    status: 'DRAFT',
    sentToContact: false,
    expectedPaymentDate: undefined,
    plannedPaymentDate: undefined,
    cISDeduction: undefined,
    cISRate: undefined,
    subTotal: 10,
    totalTax: 0,
    total: 10,
    totalDiscount: undefined,
    invoiceID: 'a6a5f183-850a-4e65-84f1-819bc040be6d',
    repeatingInvoiceID: undefined,
    hasAttachments: undefined,
    isDiscounted: false,
    payments: undefined,
    prepayments: [],
    overpayments: [],
    amountDue: 10,
    amountPaid: 0,
    fullyPaidOnDate: undefined,
    amountCredited: undefined,
    updatedDateUTC: 2021-09-24T17:53:08.470Z,
    creditNotes: undefined,
    attachments: undefined,
    hasErrors: false,
    statusAttributeString: undefined,
    validationErrors: undefined,
    warnings: undefined
  }
]
serializer call:  {
  Contacts: [
    {
      ContactID: undefined,
      ContactNumber: undefined,
      AccountNumber: undefined,
      ContactStatus: undefined,
      Name: 'Bruce Banner',
      FirstName: undefined,
      LastName: undefined,
      EmailAddress: 'hulk@avengers.com',
      SkypeUserName: undefined,
      ContactPersons: undefined,
      BankAccountDetails: undefined,
      TaxNumber: undefined,
      AccountsReceivableTaxType: undefined,
      AccountsPayableTaxType: undefined,
      Addresses: undefined,
      Phones: [Array],
      IsSupplier: undefined,
      IsCustomer: undefined,
      DefaultCurrency: undefined,
      XeroNetworkKey: undefined,
      SalesDefaultAccountCode: undefined,
      PurchasesDefaultAccountCode: undefined,
      SalesTrackingCategories: undefined,
      PurchasesTrackingCategories: undefined,
      TrackingCategoryName: undefined,
      TrackingCategoryOption: undefined,
      PaymentTerms: undefined,
      UpdatedDateUTC: undefined,
      ContactGroups: undefined,
      Website: undefined,
      BrandingTheme: undefined,
      BatchPayments: undefined,
      Discount: undefined,
      Balances: undefined,
      Attachments: undefined,
      HasAttachments: undefined,
      ValidationErrors: undefined,
      HasValidationErrors: undefined,
      StatusAttributeString: undefined
    }
  ]
}
localVarRequestOptions:  {
  method: 'PUT',
  qs: {},
  headers: {
    'user-agent': 'xero-node-4.15.0',
    'xero-tenant-id': '07f32319-daaf-4548-b589-d6f06f7b407f'
  },
  uri: 'https://api.xero.com/api.xro/2.0/Contacts',
  useQuerystring: false,
  json: true,
  body: { Contacts: [ [Object] ] }
}
contacts:  [
  Contact {
    contactID: '4e0ad2b8-f80c-42e6-bd1f-0a910b342317',
    contactNumber: undefined,
    accountNumber: undefined,
    contactStatus: 'ACTIVE',
    name: 'Bruce Banner',
    firstName: undefined,
    lastName: undefined,
    emailAddress: 'hulk@avengers.com',
    skypeUserName: undefined,
    contactPersons: [],
    bankAccountDetails: '',
    taxNumber: undefined,
    accountsReceivableTaxType: undefined,
    accountsPayableTaxType: undefined,
    addresses: [ [Address], [Address] ],
    phones: [ [Phone], [Phone], [Phone], [Phone] ],
    isSupplier: false,
    isCustomer: false,
    defaultCurrency: undefined,
    xeroNetworkKey: undefined,
    salesDefaultAccountCode: undefined,
    purchasesDefaultAccountCode: undefined,
    salesTrackingCategories: [],
    purchasesTrackingCategories: [],
    trackingCategoryName: undefined,
    trackingCategoryOption: undefined,
    paymentTerms: undefined,
    updatedDateUTC: 2021-09-24T17:54:19.280Z,
    contactGroups: [],
    website: undefined,
    brandingTheme: undefined,
    batchPayments: undefined,
    discount: undefined,
    balances: undefined,
    attachments: undefined,
    hasAttachments: undefined,
    validationErrors: undefined,
    hasValidationErrors: false,
    statusAttributeString: undefined
  }
]
```

I understand your frustration and I'm keen to help get you unstuck. If you could, please try to replicate your issue using one of the two above sample repos and submit a PR to the repo

**Comment by thrivedigital** [comment]
@RettBehrens I appreciate the followup, I forked the non-TypeScript test repo, replaced the contact object with my own data, and it worked. However, copying the exact same code from the working index.js into my own Express app, it is still sending 5 phantom contacts along with the original and erroring out.

I swapped this the createContacts method call out for an axios post request (using the readTokenSet method to grab the current access token, very helpful!) and passed the exact same contacts array with the same single contact object and it works with no issues.

Not sure at this point what the problem with the SDK is but I'm going to use this hybrid approach for now given the tight timeline for the Oauth switchover.

**Comment by MattBidewell** [comment]
@thrivedigital This is a good shout, I think we will have to implement something similar as well.

I've been experimenting with replicating it outside of my project's core code but have been unable to, which has concluded that it must be environment/project related. If I find out anymore I'll share.

**Comment by RettBehrens** [comment]
@MattBidewell @thrivedigital if you're willing, please post your `package.json` files with any sensitive info redacted. I'm wondering if there is a package besides Xero-Node you're both using that is somehow affecting the behavior since you're unable to replicate outside of your project's core code.

**Comment by sksankarraj** [comment]
Hi @RettBehrens I have tried to replicate the error yesterday using https://github.com/XeroAPI/xero-node-serializer-test, but its working fine as expected.

I have came to same conclusion as you are it should some other package which causing this issue.

here is my deps,
```
 "dependencies": {
    "@crystaldelta/json2xls": "^0.1.3",
    "aws-sdk": "^2.608.0",
    "body-parser": "^1.19.0",
    "cors": "^2.8.5",
    "dotenv": "^8.2.0",
    "elastic-apm-node": "^3.19.0",
    "express": "~4.16.1",
    "moment": "^2.29.1",
    "morgan": "~1.9.1",
    "multer": "^1.4.2",
    "multer-s3": "^2.9.0",
    "mysql2": "^2.1.0",
    "node-cron": "^2.0.3",
    "node-fetch": "^2.6.0",
    "nodemailer": "^6.6.3",
    "passport": "^0.4.1",
    "passport-jwt": "^4.0.0",
    "passport-local": "^1.0.0",
    "sequelize": "^5.21.3",
    "simple-excel-to-json": "^2.2.1",
    "winston": "^3.2.1",
    "xero-node": "^4.15.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.5.5",
    "@babel/core": "^7.5.5",
    "@babel/node": "^7.5.5",
    "@babel/preset-env": "^7.5.5",
    "@babel/register": "^7.5.5",
    "@crystaldelta/sequelize-auto-migrations": "^1.1.3",
    "babel-preset-env": "^1.7.0",
    "chai": "^4.2.0",
    "chai-http": "^4.3.0",
    "csv-stringify": "^5.5.1",
    "eslint": "^6.7.2",
    "husky": "^3.1.0",
    "lint-staged": "^9.5.0",
    "mocha": "^6.2.2",
    "nodemon": "^2.0.2",
    "nyc": "^14.1.1",
    "sequelize-cli": "^5.5.1",
    "xlsx": "^0.16.6"
  }
```

I have my doubts on Babel, but I haven't debugged it completely. Will be debugging it today.

**Comment by marlonbernardes** [comment]
This issue is caused when additional properties have been added to `Array.prototype`. More info here: #590

**Comment by robindunkerley** [comment]
Related to what I have detailed in this reply https://github.com/XeroAPI/xero-node/issues/509#issuecomment-1267358206

I have discovered that passing createContacts a contact which does not include the `phones` field will cause a validationException error to occur. This is not intuitive because the `phones` field does not appear to be a required field on type `Contact`. Interestingly it is possible using a regular http request to create a contact without the `phone` field.
